### PR TITLE
Handle more edge cases for HTTP-01 support in Nginx

### PR DESCRIPTION
**Pull request by ohemorange**
Fix nginx default checking so only checks by port.
If we still can't find a default (or found too many), make a dummy block.
Don't change the behavior for anything that's not HTTP01.

Part of #5409.

### CODE REVIEW

**Comment by ohemorange** [comment]
Things I've tested manually (making sure that location is added where it belong not added where it doesn't belong, including in nginx config files and the `/etc/letsencrypt/le_http_01_cert_challenge.conf` file):
- normal
- change port
- change port on default but add flag for deafult port
- change port on default
- no matches, turn off ipv6 on all ports
- multiple domains (check that both names are on cert)
- multiple domains one matching one not with default server
- multiple domains one matching one not without matching default server
- multiple domains neither has match

File: certbot-nginx/certbot_nginx/http_01.py [hunk 1]
```diff
@@ -71,30 +90,98 @@ def _add_bucket_directive(self):
...
+        validation = achall.validation(achall.account_key)
+        validation_path = self._get_validation_path(achall)
 
-    def _get_validation_path(self, achall):
-        return os.sep + os.path.join(challenges.HTTP01.URI_ROOT_PATH, achall.chall.encode("token"))
+        block.extend([['server_name', ' ', achall.domain],
+                      [['location', ' ', '=', ' ', validation_path],
+                        [['default_type', ' ', 'text/plain'],
+                         ['return', ' ', '200', ' ', validation]]]])
+        # TODO: do we want to return something else if they otherwise access this block?
```
Comment (bmw) [inline certbot-nginx/certbot_nginx/http_01.py]: What's nginx's default behavior here?

**Comment by bmw** [comment]
I didn't do any testing of this but I read through the code and it looks really good. Again, you're the best for turning this around so quickly.

While reviewing, I thought of a few things that are a corner cases of corner cases so it's probably not worth investing much (or maybe any) engineering effort in, but I think in theory we could do better with the address we use on the `listen` statement in some cases, mainly when creating our own dummy server block.

One case is Certbot takes an `--http-01-address` flag that matches the port flag. If someone set this, we should ideally respect it and when creating a dummy server block, we should have nginx use that address rather than listening on all interfaces. It's worth noting I didn't do this right in Apache either.

The second case is what if someone sets this and we find an existing server block to use. In theory, I think the behavior would be to use the server block only if it's listening on the same address, but this would likely get complicated when trying to compare the different variations of "listen on all interfaces".

The final case is when searching the config to see if there any IPv6 listens, we ignore (to different degrees) the address and port. We completely ignore the address and for the port, we're only looking for `ipv6only`. Even if nginx isn't listening with IPv6 on that port, we'll still use IPv6 here because it's listening on IPv6 for some address/port combination.

This is not likely to be a problem for (m)any Certbot users, but some people with valid setups have specific requirements for what addresses and ports a services should be listening to. For example, they may have multiple IP addresses allocated to a box with different services running on each. Certbot configuring nginx to use additional addresses/ports beyond what is currently in the nginx config or specified by the user can cause the challenge to fail. (It also will likely make the challenge succeed in cases it wouldn't previously.)

I think it's probably worth resolving the `--http-01-address` case for the dummy server block. Not so sure about the other cases.

**Comment by ohemorange** [comment]
Thanks for the review! I'd like our guiding star to be "are we broken for users that we would have analogously have been able to handle were we using tls-sni-01".

File: certbot-nginx/certbot_nginx/http_01.py (hunk 1 above)
Comment (ohemorange) [reply certbot-nginx/certbot_nginx/http_01.py]: 404

**Comment by bmw** [comment]
While I think my long post above about listen addresses discuss real problems people may hit, since we've never handled it right in TLS-SNI for nginx (and never got any complaints about it), I don't think we should block shipping HTTP-01 support on these features.

**Pull request merged**